contracts/modules/capital/CowSwapOperator.sol

```diff
@@ -0,0 +1,240 @@
+  bytes32 public immutable domainSeparator;
+
+  // Constants
+  address public constant ETH = 0xEeeeeEeeeEeEeeEeEeEeeEEEeeeeEeeeeeeeEEeE;
+  uint16 constant MAX_SLIPPAGE_DENOMINATOR = 10000;
+  uint256 constant MIN_VALID_TO_PERIOD = 600; // 10 minutes
+  uint256 constant MAX_VALID_TO_PERIOD = 3600; // 60 minutes
+  uint256 constant MIN_SELL_AMT_TO_FEE_RATIO = 100; // Sell amount at least 100x fee amount
+
+  event OrderPlaced(GPv2Order.Data order);
+  event OrderClosed(GPv2Order.Data order, uint256 filledAmount);
+
+  modifier onlyController() {
+    require(msg.sender == swapController, "SwapOp: only controller can execute");
+    _;
+  }
+
+  constructor(
+    address _cowSettlement,
+    address _swapController,
+    address _master,
+    address _weth,
+    address _priceFeedOracle
+  ) {
+    cowSettlement = ICowSettlement(_cowSettlement);
+    cowVaultRelayer = cowSettlement.vaultRelayer();
+    master = INXMMaster(_master);
+    swapController = _swapController;
+    weth = IWeth(_weth);
+    priceFeedOracle = IPriceFeedOracle(_priceFeedOracle);
+    domainSeparator = cowSettlement.domainSeparator();
+  }
+
+  receive() external payable {}
+
+  function getDigest(GPv2Order.Data calldata order) public view returns (bytes32) {
+    bytes32 hash = GPv2Order.hash(order, domainSeparator);
+    return hash;
+  }
+
+  function getUID(GPv2Order.Data calldata order) public view returns (bytes memory) {
+    bytes memory uid = new bytes(56);
+    bytes32 digest = getDigest(order);
+    GPv2Order.packOrderUidParams(uid, digest, order.receiver, order.validTo);
+    return uid;
+  }
+
+  function placeOrder(GPv2Order.Data calldata order, bytes calldata orderUID) public onlyController {
+    // Validate there's no current order going on
+    require(currentOrderUID.length == 0, "SwapOp: an order is already in place");
+
+    // Order UID verification
+    validateUID(order, orderUID);
+
+    // Validate basic CoW params
+    validateBasicCowParams(order);
+
+    // Validate feeAmount is not too high
+    require(order.sellAmount / order.feeAmount >= MIN_SELL_AMT_TO_FEE_RATIO, "SwapOp: Fee is above 1% of sellAmount");
+
+    // Local variables
+    IPool pool = _pool();
+    uint256 totalOutAmount = order.sellAmount + order.feeAmount;
+
+    if (isSellingEth(order)) {
+      // Validate min/max setup for buyToken
+      IPool.SwapDetails memory swapDetails = pool.getAssetSwapDetails(address(order.buyToken));
+      require(swapDetails.minAmount != 0 || swapDetails.maxAmount != 0, "SwapOp: buyToken is not enabled");
+      uint256 buyTokenBalance = order.buyToken.balanceOf(address(pool));
+      require(buyTokenBalance < swapDetails.minAmount, "SwapOp: can only buy asset when < minAmount");
+      require(buyTokenBalance + order.buyAmount <= swapDetails.maxAmount, "SwapOp: swap brings buyToken above max");
+      require(buyTokenBalance + order.buyAmount >= swapDetails.minAmount, "SwapOp: swap leaves buyToken below min");
```

Comment:
I just checked the logic from the old SwapOperator and you are right, when exchanging eth for an asset it's not required that the final amount is > min. I think my rationale for this was to avoid the possibility of a compromised swap controller making a lot of small trades to burn fees (since I removed the swap frequency validation). Your argument is valid so I think we should remove that particular restriction. Do you think we can ignore the potential scenario of a bad swap controller executing a lot of small swaps? or should I add the frequency check to prevent it ?

Reply:
As per our discussion, let's implement a trade frequency check.